internal/access/longpolling.go

```diff
@@ -0,0 +1,67 @@
+package access
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/infrahq/infra/internal"
+	"github.com/infrahq/infra/internal/logging"
+	"github.com/infrahq/infra/internal/server/data"
+)
+
+type updateIndexable interface {
+	UpdateIndex() int64
+	ItemCount() int
+}
+
+func blockingRequest[Result updateIndexable](rCtx RequestContext, listenOpts data.ListenForNotifyOptions, query func() (Result, error), lastUpdateIndex int64) (Result, error) {
+
+	listener, err := data.ListenForNotify(rCtx.Request.Context(), rCtx.DataDB, listenOpts)
+	if err != nil {
+		// nolint: gocritic
+		return *new(Result), fmt.Errorf("listen for notify: %w", err)
+	}
+	defer func() {
+		// use a context with a separate deadline so that we still release
+		// when the request timeout is reached
+		ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+		defer cancel()
+		if err := listener.Release(ctx); err != nil {
+			logging.L.Error().Err(err).Msg("failed to release listener conn")
+		}
+	}()
+
+	result, err := query()
+	if err != nil {
+		return result, err
+	}
+
+	// The query returned results that are new to the client
+	if result.ItemCount() > 0 && result.UpdateIndex() > lastUpdateIndex {
```

Comment:
This `result.ItemCount` is new, and that seems like it introduces a bug. If all the grants are deleted, the client needs to know the response is now empty, so a response should be sent when `itemCount == 0`. 

I guess we don't have test coverage of that scenario.

Why is this new clause necessary?

Reply:
When there are no records in the table and this guard is not in place, this causes the function to return early with no results instead of waiting. This would force the client to make another request. Returning with zero records doesn't seem to be that useful for a function that's supposed to block until _new_ data arrives.

Reply:
> When there are no records in the table and this guard is not in place, this causes the function to return early with no results instead of waiting

That's the correct behaviour. The first query should always return the state of the world so that the client knows it has the current state. The next query that comes in should block properly. Is the second query not blocking? That would be a bug.

> Returning with zero records doesn't seem to be that useful for a function that's supposed to block until new data arrives.

The zero state is a completely valid state. When a client receives this first empty request it goes from "unknown" state to a valid state with no data. That **MUST** be returned to the client for the system to behave correctly.

Reply:
> "No data" can also be new data (when everything is deleted).

State of the world is not useful to me; I only need pending, unanswered requests, and there are often none. I believe without this check the client thrashes, and continually hits the server immediately over and over again, so we need a solution to that before I revert this change.

Reply:
The update index is what should prevent that. It sounds like maybe you aren't using an update index, which is maybe the problem.

Reply:
The update index doesn't have to come from the DB. You could fake it in code if you can derive it from the state of the query.

If all your concerned about is "has this been updated", then you could use the following for update index:

```
1 - pending
2 - done
```